### ISSUE TITLE
Simulations aren't working

### ISSUE THREAD

**Issue by russbramblett** [xuzs9298/CGEpy#12]
I believe I've modified the code such that P0 increases by 2%. This change doesn't have an affect on the output, but it should. 

Changes where made to lines 365, 584, and 882,

**Comment by russbramblett** [comment]
It looks like the problem is that the minimize function is exiting prior to reaching the optimal solution. When I add the maxiter options to the minimize function,

`results = minimize(objFun, initialVals, method='SLSQP', #jac = False, 
                   bounds = None, constraints=cons, tol = toler, 
                   options = {'maxiter':len(initialVals)*2})`

the optimizer never iterates to completion. I let the code run overnight and it still didn't finish. 

Reading up on this, the solution might be to include the Jacobian of the objective function in the minimizer. This is non-trivial as it will require a function to calculate the derivative at each value.

**Comment by xuzs9298** [comment]
It looks like one of the problems is that: after calibration, the equations are not satisfied. It's caused by the eval() function. I believe I've solved this one. 

But the variable still won't converge on the given initial values, it will change within the given bounds and get a better solution. I'm still looking into it. Hopefully within several days.

**Comment by russbramblett** [comment]
Hopefully you can get it because calculating the Jacobians probably isn't a viable option. I calculated a few and it is not a good time.

**Comment by xuzs9298** [comment]
Hi Russ,

In the latest commit, I uploaded what I have right now. 

Some major changes I made:

1. There's now a method avl.getLabel(index) that returns the label for a given index. It's easy for the analysis.
2. Instead of creating a virtual function with eval(), all the equations writes to a constraint.py now. And this constraint.py is executed before the minimize method.
3. I added the bounds as it is in GAMS(line 1460-1485), **some variables are fixed in GAMS**, that parts corresponds to line 1440-1457.
4. line 1488, 1489 is fixing HH, which is extra from GAMS, only with more constraint can we successfully calibrate it to the initial values. (As I said in the email, HW also works). I was trying to use this to track down which equation is wrong, but it doesn't work out.
5. The difference after the optimization is now outputted to a file. You can change the file name in line 1545.

Some minor changes/mistakes I fixed:
1. value of table A was wrong, fixed.
2. equation POPEQ has a parenthesis issue, fixed.
3. TAUFX_SUM, fixed.
4. Now the program checks for each constraint before and after the optimization.
5. If you run the program with the bounds it should have(no extra HH bounds). You will get  a "positive directional derivative for linesearch" after about 60 iterations. It will show "success: false", the constraints is not satisfied after the optimization, but close. The optimizer is not sure whether this is the optimal solution. The spi is increasing for a significant amount, so there's still something wrong in the equations.

**Comment by xuzs9298** [comment]
Here's what I think we should do next:
I'll check again for each variable and each table used in equations. It's possible that the values are wrong but the equation is satisfied with initial values, just like some mistakes I fixed previously.

You can run the program without the HH bounds, and in the result it returned, check results.x for each equation and see which equation is in the wrong format. (results.x doesn't satisfy all constraints, but should be within 0.1). Since you know those equations better than I am. Maybe you can spot it much easier.

**Comment by russbramblett** [comment]
Sounds like a good plan. I'll get to work on it.

-Russ

**Comment by russbramblett** [comment]
Did you get it figured out?

**Comment by xuzs9298** [comment]
Yes. I'm just about to verify it again and upload it.

**Comment by xuzs9298** [comment]
I also tried P0 = 1.02 and it gets the same output with GAMS.

**Comment by russbramblett** [comment]
Nice. I am sitting here waiting for the thing to optimize.

I'll look for your commit.

-Russ

**Comment by xuzs9298** [comment]
I just committed it. I don't have github on my desktop, so I just copied the file to my laptop and uploaded it.

Method I used to debug it:
1. I run the optimizer a second time with almost fixed HW, so the optimizer will produce a wrong but better solution.
2. I run GAMSDebug.py to generate all the value assignment statements in GAMS.
3. Then I compare each equation with GAMS equation. I will also upload that GAMS file later.
